clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -493,8 +493,9 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
         // TODO i#5949: For WOW64 instr_is_syscall() always returns false here as it
         // tries to check adjacent instrs; we disable this check until that is solved.
 #if !defined(WINDOWS) || defined(X64)
-        if (shard->prev_instr_.decoding.has_valid_decoding) {
-            report_if_false(shard, shard->prev_instr_.decoding.is_syscall,
+        if (shard->prev_instr_.decoding != nullptr &&
```

Comment:
There are now nullptr checks everywhere: including a number that were overlooked by this PR, making the point that this is now harder to use.  For invariant checker I guess it is valid to not have any decode info (legacy trace w/o encodings: this tool doesn't support the old mapping); but this change to now have to check for nullptr everywhere doesn't feel as user-friendly and feels error-prone.  Should we consider changing it around to have the base decode-info class always there with a `bool decoding_exists` and `bool decoding_valid`?  And the user doesn't subclass it but creates an object stored inside it: I guess it has to be the one templated then.  Hmm.

Reply:
I think from the standpoint of instr_decode_cache_t, it makes sense to return nullptr if no decoding info is available.

It seems that the invariant checker was relying on decoding_info_t defaults to simplify many of these conditions. Probably worth an audit to ensure that the conditions get evaluated as intended when a default decoding_info_t is found.

In any case IMHO a good way forward is to let prev_instr_.decoding continue be a non-ptr object, and set it to the default-constructed object if the decode cache returned nullptr. Would require making a copy of the decoding_info_t in the invariant checker but that should be ok.